pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,331 @@
+		// different authoritative owner for a partition, then each time cluster metadata is updated the Kafka client
+		// will client a new connection for each partition, leading to an high connections churn rate.
+		kgo.MetadataMaxAge(10*time.Second),
+		kgo.WithHooks(r.metrics.kprom),
+		kgo.WithLogger(newKafkaLogger(r.logger)),
+	)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "creating kafka client")
+	}
+
+	return client, nil
+}
+
+func (r *PartitionReader) fetchLastCommittedOffsetWithRetries(ctx context.Context) (offset kgo.Offset, err error) {
+	var (
+		retry = backoff.New(ctx, backoff.Config{
+			MinBackoff: 100 * time.Millisecond,
+			MaxBackoff: 2 * time.Second,
+			MaxRetries: 10,
+		})
+	)
+
+	for retry.Ongoing() {
+		offset, err = r.fetchLastCommittedOffset(ctx)
+		if err == nil {
+			return offset, nil
+		}
+
+		level.Warn(r.logger).Log("msg", "failed to fetch last committed offset", "partition", r.partitionID, "err", err)
+		retry.Wait()
+	}
+
+	// Handle the case the context was canceled before the first attempt.
+	if err == nil {
+		err = retry.Err()
+	}
+
+	return offset, err
+}
+
+func (r *PartitionReader) fetchLastCommittedOffset(ctx context.Context) (kgo.Offset, error) {
+	cl, err := kgo.NewClient(kgo.SeedBrokers(r.kafkaAddress))
+	if err != nil {
+		return kgo.NewOffset(), errors.Wrap(err, "unable to create admin client")
+	}
+	adm := kadm.NewClient(cl)
+	defer adm.Close()
+
+	offsets, err := adm.FetchOffsets(ctx, consumerGroup)
+	if errors.Is(err, kerr.UnknownTopicOrPartition) {
+		// In case we are booting up for the first time ever against this topic.
+		return kgo.NewOffset().AtEnd(), nil
+	}
+	if err != nil {
+		return kgo.NewOffset(), errors.Wrap(err, "unable to fetch group offsets")
+	}
+	offset, _ := offsets.Lookup(r.kafkaTopic, r.partitionID)
+	return kgo.NewOffset().At(offset.At), nil
+}
+
+func (r *PartitionReader) commitLoop(ctx context.Context) {
+	defer r.commitLoopWg.Done()
+
+	commitTicker := time.NewTicker(r.commitInterval)
+	defer commitTicker.Stop()
+
+	toCommit := kadm.Offsets{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case fetches := <-r.commitFetches:
+			fetches.EachPartition(func(p kgo.FetchTopicPartition) {
+				if len(p.Records) == 0 {
+					return
+				}
+				r := p.Records[len(p.Records)-1]
+				toCommit.AddOffset(r.Topic, r.Partition, r.Offset+1, r.LeaderEpoch)
+			})
+		case <-commitTicker.C:
+			if len(toCommit) == 0 {
+				continue
+			}
+			committed, err := r.admClient.CommitOffsets(ctx, consumerGroup, toCommit)
+			if err != nil || !committed.Ok() {
+				level.Error(r.logger).Log("msg", "encountered error while committing offsets", "err", err, "commit_err", committed.Error())
+			} else {
+				committedOffset, _ := committed.Lookup(r.kafkaTopic, r.partitionID)
+				level.Debug(r.logger).Log("msg", "committed offset", "offset", committedOffset.Offset.At)
+				clear(toCommit)
+			}
+		}
+	}
+}
+
+type readerMetrics struct {
```

Comment:
[non blocking] Instead of having a shared struct between `PartitionReader` and `pusherConsumer`, have you considered removing this struct at all and having each component defining their own metrics, so passing the registerer to `newPusherConsumer()`.

Reply:
i like how the struct declitters the ParitionReader. I changes pusherConsumer to instantiate its own metrics, so now the struct isn't shared at least